src/binary.rs

```diff
@@ -0,0 +1,322 @@
+    let text = text.trim();
+
+    let system_version = match app {
+        Application::WasmBindgen => text.splitn(2, ' ').nth(1).context("missing version")?.to_owned(),
+        Application::WasmOpt => text.splitn(2, ' ').nth(1).context("missing version")?.replace(' ', "_"),
+    };
+
+    if system_version == version {
+        Ok(path)
+    } else {
+        bail!("not found")
+    }
+}
+
+/// Check whether a given path exists, is a file and marked as executable (unix only).
+fn is_executable(path: &Path) -> Result<bool> {
+    if !path.exists() {
+        return Ok(false);
+    }
+
+    let metadata = path.metadata().with_context(|| anyhow!("failed getting metadata of {:?}", path))?;
+    if !metadata.is_file() {
+        return Ok(false);
+    }
+
+    #[cfg(unix)]
+    {
+        use std::os::unix::fs::PermissionsExt;
+        if metadata.permissions().mode() & 0o100 == 0 {
+            return Ok(false);
+        }
+    }
+
+    Ok(true)
+}
+
+/// Download a file from its remote location in the given version, extract it and make it ready for
+// execution at the given location.
+#[tracing::instrument(level = "trace")]
+async fn download(app: Application, version: &str) -> Result<PathBuf> {
+    tracing::info!(version = version, "downloading {}", app.name());
+
+    let cache_dir = cache_dir().context("failed getting the cache directory")?;
+    let client = surf::client().with(Redirect::new(5));
+
+    let mut resp = client
+        .get(app.url(version)?)
+        .send()
+        .await
+        .map_err(|e| e.into_inner())
+        .context("error sending HTTP request")?;
+    ensure!(
+        resp.status().is_success(),
+        "error downloading binary file: {:?}\n{}",
+        resp.status(),
+        app.url(version)?
+    );
+
+    let temp_out = cache_dir.join(format!("{}-{}.tmp", app.name(), version));
+    let mut file = AsyncFile::create(&temp_out).await.context("failed creating temporary output file")?;
+
+    io::copy(resp.take_body().into_reader(), &mut file)
+        .await
+        .context("failed downloading the archive")?;
+
+    Ok(temp_out)
+}
+
+/// Install an application from a downloaded archive locating and copying it to the given target
+/// location.
+#[tracing::instrument(level = "trace")]
+fn install(app: Application, archive_file: &mut File, target: &Path) -> Result<()> {
+    tracing::info!("installing {}", app.name());
+
+    let mut archive = TarArchive::new(GzDecoder::new(archive_file));
+    let mut file = extract_file(&mut archive, target, Path::new(app.path()))?;
+
+    set_executable_flag(&mut file)?;
+
+    for path in app.extra_paths() {
+        // archive must be opened for each entry as tar files don't allow jumping forth and back.
+        let archive_file = archive.into_inner().into_inner();
+        archive_file.seek(SeekFrom::Start(0))?;
+
+        archive = TarArchive::new(GzDecoder::new(archive_file));
+        extract_file(&mut archive, target, Path::new(path))?;
+    }
+
+    Ok(())
+}
```

Comment:
Generally, I would say that these should also be changed into `Application` methods (basically only need to change the signature and `s/app/self/`).

Reply:
I didn't want to tie too much logic to the `Application` enum as it only needs some names and paths in this functions and these details could technically be replaced by 1 or 2 extra parameters so the implementation is rather flexible.

Just trying to not make this implementation too OOP style 😅 

The overall design is that the `binary` module contains most of the logic and the `Application` just provides some metadata to fulfill the task like download URL, paths in the downloaded tarballs and so on. This way it's not to tightly bound to the `Application` and things are independent where it makes sense. This can help in writing unit tests later as well.

What do you think?

Reply:
That's a tough one. I've definitely gone back and forth quite a lot in my mind about these issues over the years. These days, I do think it is nice when it is possible to encapsulate type specific functionality as a set of methods. However, TBH, it really is not going to make a big difference at this point. Let's keep this code as is for now! Thanks for taking the time to defend your position `:)`